### ISSUE TITLE
[Feature Request] Test filters from command line by replaying request or response

### ISSUE THREAD

**Issue by innodonni** [e2guardian/e2guardian#454]
I would like to be able to debug the following issues with false positives and false negatives:
a) content regexp not matching content
b) weighted phrase not matching content
c) denied content that is not reproducible from the logged url (UA, cookie or dynamic page related)
I propose the following features would solve these issues:
1. Command line argument `--df FILTERGROUP` to apply a test to a given filter group
2. Command line argument `--dh -` to test filtering rules against content from stdin (including headers)
3. Command line argument `--di -` to test filtering rules against content from stdin (excluding headers)
4. Command line argument `--du URL` to test filtering rules against a URL (with/without `--di`)
5. Command line argument `--dl LEVELS` to apply debug levels to stderr for a test
6. Option to log banned requests with responses for a filter group for replaying later on.

To avoid reloading lists for each test, these would probably require a running daemon to spawn a thread with standard in/out piped to the test process somehow.

The difficulty in implementing this change myself is:
i) The HTTPHeader::in and ::out methods presuming a Socket rather than BaseSocket.
ii) How to make the fat controller listen on pipes from test processes and how to send the content from stdin through to the daemon
iii) How to make the ConnectionHandler log to (piped) stderr for tests

I have created this issue to discuss the merits of this solution along with alternatives and tips. This should also help with #264 .

**Comment by fredbcode** [comment]
Hi innodonni

Maybe we could just use the newest debug mode for this (v5.2 and 
v5develop) with some new tags, and add a new "learning mode" this 
feature could trace without really deny (just logging).

Together with debug mode this can be powerful


https://github.com/e2guardian/e2guardian/blob/v5.2/configs/e2guardian.conf.in#L20

**Comment by innodonni** [comment]
Yes, I realise logheadervalue="cookie:" and loguseragent=on should allow more accurately replaying requests.
To replay responses, I could write a short script to pipe a response with headers to netcat listening on a local port, and then request it through the proxy again.
I think I could modify the storyboard to test output without blocking but it wouldn't allow me to unobtrusively monitor blocked responses for end-users. If content modification were a storyboard function alongside or called within checkresponse, debug functions could be inserted to log the response both before and after if desired.
If controlled only by debuglevel, the remaining issue is it could do with a similar option to loglevel, or separate tags to minimize IO e.g. REQBLOCKED, REQALLOWED, RESPBLOCKED, RESPALLOWED.
I would be happy to help converting some more DGDEBUG to debug categories as it should be a much simpler job. I've got several ideas to update and improve docs and config file comments too.

**Comment by innodonni** [comment]
Here's an idea: if the user resubmits the request with a certain cookie, then the server could check that the request is still denied, send the response as an attachment by email to the server administrator and allow the administrator to bypass, add exception or type a new rule to add from a separate web interface. That would be lower volume, use no disk space and also reuse several existing pieces of code: a) CGI script for denied page b) email notification c) bypass cookie

**Comment by fredbcode** [comment]
Thank you,

In reality this is more a less what happen with BYPASS (cookie) here I'm 
using something like this, the user is redirected to a PHP page and the 
web site sends an email to admin, after the url is allowed/denied 
definitely with a web interface by admin and the new  blacklist is 
pushed to E2

Works well but actually I'm seeing two bad "points"

1) The parameters are in URL request, so the user can change the user 
name, IP, etc

2) E2guardian must reload the full blacklist


We must fix this two issues before update the BYPASS mode, but I'm not 
sure of how ?


Ideas: for 1 a second cookie with the parameters inside, less easier to 
change but not perfect and for 2 a REST module for e2guardian like this 
it can wait order from outside (like add this domain to blacklist, 
reload, remove domain from whitlist, etc)

About first point it must be transparent from user point of view


Fred

**Comment by fredbcode** [comment]
However, I digress ...
About " if the user resubmits the request with a certain cookie" maybe it's s a bit complicated for admin ?
But perhaps there is no easy way, I don't know 
And what do you think to tag a testing list http://contentfilter.futuragts.com/wiki/doku.php?id=downloadable_blacklists

Something like 
#listcategory: "banned (test)"

When there is "(test)" tag the request is logged with more description but not blocked

**Comment by innodonni** [comment]
I am thinking a procedure secure enough to run on a production server and
easy and flexible enough for people to use out of the box.
Complexity is for us developers to deal with :^)
I think I see where you are going. From a user perspective, they see the
block page and click to ask the admin to unblock. From e2guardian's
perspective, you want all requests to the proxy port to be real ones, not
requests to unblock - but bypass doesn't follow that rule. You might say if
I need a server-side script, I might as well make it do its job internally,
including download the requested page.
There is the issue I am trying to solve - e2guardian knows how and why the
content was banned, not the web server. Even if it requests the page again,
it may get different content that does not trigger the same rules - and the
content it does store cannot be replayed through e2guardian.
In addition, it would be useful to other admins to include the solution in
the e2guardian distribution - if not in the compiled software itself. I am
proposing to write whatever component is needed, but need to know how it
should best be done.

Regarding tagging lists, any feature to switch on could be either included
on a separate line
#listcategory: "allowed but monitored"
#listaction: "logcontent"
#listname: "logcontent"
...still, safer to put the messageno and explicitly reference the lists
instead of storing all the list metadata like that,
or coded in the storyboard using the existing log lists:
if(fullurlin,log) logcontent

Or even, in my use case, something like
if(cookiein,pleaseunblock) savecachedcontent

I think I should create a separate issue for the "log content when unblock
requested" idea, not to confuse matters, especially since many people will
already have existing web scripts and the bypass feature may be enough for
most people. The main problem - replaying the exact response content the
user was accessing - is the main issue here.

I think it may be simplest, now I think about it, to create a second
executable, /usr/sbin/e2guardianctl for example, whose main purpose would
be to add rules both to the file and running daemon at the same time.
However, it could also allow running content through the same rules as the
server.
The root issue seems not to be how to do this, but how to load the rules
fast enough without needing to share memory with the server process.
Perhaps the best thing would be to write the code to replay
requests/responses using rules loaded from disk, and then change the
loading code for both to use a faster cache/binary blobs (squidguard does
something like this) and lazy rebuilding just like GNU Make.
Sounds like a useful project either way. :-)
